src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -1834,6 +1838,26 @@ static void post_deoptimization_event(CompiledMethod* nm,
 
 #endif // INCLUDE_JFR
 
+void Deoptimization::print_ul(CompiledMethod* nm, Method* tm, intptr_t pc, frame& fr, int trap_bci,
```

Comment:
From what I can see, this function only called in this cpp file so it only needs to be defined only in this cpp file.  It doesn't seem to use any Deoptimization class members so doesn't need to be a class member.  Then you can revert changes to deoptimize.hpp.

Reply:
And can you call it something like log_deopt(), since the "print" function is generally a function that just prints to an outputStream without including logic about logging.

Reply:
Sounds good, fixed.